ddtrace/profiling/collector/_memalloc_tb.cpp

```diff
@@ -121,35 +233,103 @@ traceback_t::traceback_t(void* ptr,
+    sample.push_heap(size);
+
+    // Get thread native_id and name from Python's _threading module and push to sample
+    push_threadinfo_to_sample(sample, thread_id);
+
+    // Collect frames from the Python frame chain and push to Sample
+    // We push frames as we collect them (root to leaf order).
+    // Note: Sample.push_frame() comment says it "Assumes frames are pushed in leaf-order",
+    // but we push root-to-leaf. Set reverse_locations so the sample will be reversed when exported.
+    sample.set_reverse_locations(true);
+    for (PyFrameObject* frame = pyframe; frame != NULL;) {
         // TODO(dsn): add a truncated frame to the traceback if we exceed the max_nframe
-        total_nframe++;
+
+        // Extract frame info for Sample
+        int lineno_val = PyFrame_GetLineNumber(frame);
+        if (lineno_val < 0)
+            lineno_val = 0;
 
 #ifdef _PY39_AND_LATER
-        PyFrameObject* back = PyFrame_GetBack(pyframe);
-        Py_DECREF(pyframe);
-        pyframe = back;
+        PyCodeObject* code = PyFrame_GetCode(frame);
 #else
-        pyframe = pyframe->f_back;
+        PyCodeObject* code = frame->f_code;
+#endif
+
+        // Extract frame info for Sample
+        // Create string_views directly from bytes buffers, push_frame copies them immediately,
+        // then we can safely DECREF the bytes objects
+        std::string_view name_sv = "<unknown>";
+        std::string_view filename_sv = "<unknown>";
+
+        PyObject* name_bytes = nullptr;
+        PyObject* filename_bytes = nullptr;
+
+        if (code != NULL) {
+            if (code->co_name) {
+                name_bytes = PyUnicode_AsUTF8String(code->co_name);
+                if (name_bytes) {
+                    const char* name_ptr = PyBytes_AsString(name_bytes);
+                    if (name_ptr) {
+                        Py_ssize_t name_len = PyBytes_Size(name_bytes);
+                        name_sv = std::string_view(name_ptr, name_len);
+                    }
+                }
+            }
+
+            if (code->co_filename) {
+                filename_bytes = PyUnicode_AsUTF8String(code->co_filename);
+                if (filename_bytes) {
+                    const char* filename_ptr = PyBytes_AsString(filename_bytes);
+                    if (filename_ptr) {
+                        Py_ssize_t filename_len = PyBytes_Size(filename_bytes);
+                        filename_sv = std::string_view(filename_ptr, filename_len);
+                    }
+                }
+            }
+        }
+
+        // Push frame to Sample (root to leaf order)
+        // push_frame copies the strings immediately into its StringArena, so it's safe to
+        // DECREF the bytes objects after this call
+        sample.push_frame(name_sv, filename_sv, 0, lineno_val);
+
+        // Now safe to release the bytes objects since push_frame has copied the strings
+        Py_XDECREF(name_bytes);
+        Py_XDECREF(filename_bytes);
```

Comment:
if someone who understands python refcounting could confirm that that would be great

Reply:
[PyUnicode_AsUTF8String](https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsUTF8String) returns a new reference so when no longer needed the refcount needs to be decremented